packages-ts/gauntlet-terra-cw20-multisig/src/commands/multisig.ts

```diff
@@ -0,0 +1,250 @@
+      funds: {
+        denom: string
+        amount: string
+      }[]
+      msg: string
+    }
+  }
+}
+
+enum Action {
+  CREATE = 'create',
+  APPROVE = 'approve',
+  EXECUTE = 'execute',
+  NONE = 'none',
+}
+
+type State = {
+  threshold: number
+  nextAction: Action
+  owners: AccAddress[]
+  approvers: string[]
+  // https://github.com/CosmWasm/cw-plus/blob/82138f9484e538913f7faf78bc292fb14407aae8/packages/cw3/src/query.rs#L75
+  currentStatus?: 'pending' | 'open' | 'rejected' | 'passed' | 'executed'
+  data?: WasmMsg[]
+}
+
+export const wrapCommand = (command) => {
+  return class Multisig extends TerraCommand {
+    command: TerraCommand
+    multisig: AccAddress
+    multisigGroup: AccAddress
+
+    static id = `${command.id}:multisig`
+
+    constructor(flags, args) {
+      super(flags, args)
+
+      this.command = new command(flags, args)
+
+      if (!AccAddress.validate(process.env.MULTISIG_ADDRESS)) throw new Error(`Invalid Multisig wallet address`)
+      if (!AccAddress.validate(process.env.MULTISIG_GROUP)) throw new Error(`Invalid Multisig group address`)
+      this.multisig = process.env.MULTISIG_ADDRESS as AccAddress
+      this.multisigGroup = process.env.MULTISIG_GROUP as AccAddress
+    }
+
+    makeRawTransaction = async (signer: AccAddress, state?: State) => {
+      const message = await this.command.makeRawTransaction(this.multisig)
+
+      const operations = {
+        [Action.CREATE]: this.executePropose,
+        [Action.APPROVE]: this.executeApproval,
+        [Action.EXECUTE]: this.executeExecution,
+        [Action.NONE]: () => {
+          throw new Error('No action needed')
+        },
+      }
+
+      if (state.nextAction !== Action.CREATE) {
+        this.require(
+          await this.isSameProposal(state.data, [this.toWasmMsg(message)]),
+          'The transaction generated is different from the proposal provided',
+        )
+      }
+
+      return operations[state.nextAction](signer, Number(this.flags.proposal), message)
+    }
+
+    isSameProposal = (proposalMsgs: WasmMsg[], generatedMsgs: WasmMsg[]) => {
+      return isDeepEqual(proposalMsgs, generatedMsgs)
+    }
+
+    toWasmMsg = (message: MsgExecuteContract): WasmMsg => {
+      return {
+        wasm: {
+          execute: {
+            contract_addr: message.contract,
+            funds: message.coins.toArray().map((c) => c.toData()),
+            msg: Buffer.from(JSON.stringify(message.execute_msg)).toString('base64'),
+          },
+        },
+      }
+    }
+
+    executePropose: ProposalAction = async (signer, _, message) => {
+      logger.info('Generating data for creating new proposal')
+      const proposeInput = {
+        propose: {
+          description: command.id,
+          msgs: [this.toWasmMsg(message)],
+          title: command.id,
+          // TODO: Set expiration time
+          // latest: { never: {} }
+        },
+      }
+      return new MsgExecuteContract(signer, this.multisig, proposeInput)
```

Comment:
Where do we define the wasm function to call? e.g. `begin_proposal`?

Reply:
Not sure is this is what you mean, but here we decide which operation to call: https://github.com/smartcontractkit/chainlink-terra/pull/111/files#diff-c7c2eafb6b93bd0e456c2fd133626df8051e82da80db2ad3695f45fe6cbd72bbR88
That's based on `state.nextAction`